File: 프로그래머스/연습문제/Lv2/미로 탈출/미로 탈출.cpp
#include <string>
#include <vector>
#include <queue>
using namespace std;

int BFS(vector<string> maps, int start_i, int start_j, int lever_i, int lever_j){
    vector<vector<int>> visit;
    for(int i=0; i<maps.size(); i++){
        vector<int> v(maps[0].size(), -1);
        visit.push_back(v);
    }
    
    queue<pair<int,int>> q;
    q.push(make_pair(start_i, start_j));
    visit[start_i][start_j] = 0;
    
    int di[4] = {-1, 0, 1, 0};
    int dj[4] = {0, 1, 0, -1};
    
    while(q.size()){
        int tmp_i = q.front().first;
        int tmp_j = q.front().second;
        q.pop();
        
        if(tmp_i==lever_i && tmp_j==lever_j)
            return visit[tmp_i][tmp_j];
        
        for(int i=0; i<4; i++){
            int move_i = tmp_i+di[i];
            int move_j = tmp_j+dj[i];
            
            if(move_i<0 || move_i>=maps.size()) continue;
            if(move_j<0 || move_j>=maps[0].size()) continue;
            
            if(visit[move_i][move_j]==-1 && maps[move_i][move_j]!='X'){
                q.push(make_pair(move_i, move_j));
                visit[move_i][move_j] = visit[tmp_i][tmp_j]+1;
            }
        }
    }
    
    return visit[lever_i][lever_j];
}

int solution(vector<string> maps) {
    pair<int,int> start, lever, exit;
    for(int i=0; i<maps.size(); i++){
        for(int j=0; j<maps[0].size(); j++){
            if(maps[i][j] == 'S'){
                start.first = i;
                start.second = j;
            }
            else if(maps[i][j] == 'L'){
                lever.first = i;
                lever.second = j;
            }
            else if(maps[i][j] == 'E'){
                exit.first = i;
                exit.second = j;
            }
        }
    }
    
    int toLever = BFS(maps, start.first, start.second, lever.first, lever.second);
    int toExit = BFS(maps, lever.first, lever.second, exit.first, exit.second);
    
    if(toLever==-1 || toExit==-1) return -1;
    else return toLever+toExit;
}

File: 프로그래머스/연습문제/Lv2/미로 탈출/미로 탈출.py
def BFS(maps, start, end):
    visit = [[-1 for i in range(len(maps[0]))] for i in range(len(maps))]
    q = [[start[0], start[1]]]
    visit[start[0]][start[1]] = 0

    di = [-1, 0, 1, 0]
    dj = [0, 1, 0, -1]

    while len(q) > 0:
        tmp_i = q[0][0]
        tmp_j = q[0][1]
        q.pop(0)

        if tmp_i==end[0] and tmp_j==end[1]:
            return visit[tmp_i][tmp_j]
        
        for i in range(4):
            move_i = tmp_i+di[i]
            move_j = tmp_j+dj[i]

            if move_i<0 or move_i>=len(maps): continue
            if move_j<0 or move_j>=len(maps[0]): continue

            if visit[move_i][move_j]==-1 and maps[move_i][move_j]!='X':
                q.append([move_i, move_j])
                visit[move_i][move_j] = visit[tmp_i][tmp_j]+1

    return visit[end[0]][end[1]]

def solution(maps):
    start,lever,exit = [],[],[]
    for i in range(len(maps)):
        for j in range(len(maps[0])):
            if maps[i][j] == 'S':
                start.append(i)
                start.append(j)
            elif maps[i][j] == 'L':
                lever.append(i)
                lever.append(j)
            elif maps[i][j] == 'E':
                exit.append(i)
                exit.append(j)

    toLever = BFS(maps, start, lever)
    toExit = BFS(maps, lever, exit)

    if toLever==-1 or toExit==-1: return -1
    else: return toLever+toExit
